### DISCUSSION TITLE
Running voxnovel on k80 with CUDA 11.8

### DISCUSSION THREAD

**Discussion by dmarsh400** [DrewThomasson/VoxNovel#31] [Q&A] [answered]
I just installed a K80 and I am trying to run VoxNovel but it is telling me my driver is too old and to update or to compile CUDA from source, which I did, the K80 runs on 11.4 but I see it is compatible with 11.8

I was looking at the linux requirements and it shows CUDA 12 as needed to run VoxNovel

is this correct or is there any way I can get this to run with 11.8?

Thanks

**Comment by DrewThomasson** [comment]
## It should run fine with CUDA 11.8

#### I know this because:
As in order to fix Headless VoxNovel Google Colab I HAD to downgrade the Google Colab env to CUDA 11.8 to get it to run in the first place

Referenced Google Colab: https://colab.research.google.com/drive/15pp2hFBo2fD3legDQfWY5DMt-aI-HKKF?usp=sharing

**Comment by dmarsh400** [comment]
![Screenshot_2024-09-05_09-03-28](https://github.com/user-attachments/assets/34e69db9-79eb-4dda-a9f6-24e616ba2103)

I keep getting this error when starting up, not sure exactly how to proceed, I have been trying to get this to work and I am not sure what I am doing wrong

**Reply by DrewThomasson** [reply]
You used the single command auto-installer for ubuntu right?

**Reply by dmarsh400** [reply]
yes I did, and I also tried a reinstall on the single command line to see if that would update or fix the issue

**Reply by dmarsh400** [reply]
how would I uninstall and reinstall if I needed to do that?

**Reply by DrewThomasson** [reply]
You should just be able to just delete the VoxNovel conda env

`conda remove --name VoxNovel --all` and that'll remove all the installed pips

But like, the install would just be recreating that env and doing all the pip installs again

**Reply by DrewThomasson** [reply]
I do plan on making a Ubuntu uninstall script later to hunt down the xtts booknlp and nltk model files that were downloaded tho and remove those as well.

But the model files wouldn't be causing your issue.

**Reply by dmarsh400** [reply]
Ok I will try to remove and reinstall and see if that fixes the issue, if not I may have to start with a clean install of ubuntu

**Reply by DrewThomasson** [reply]
Wait run this command to check your cuda version

`nvcc --version`

Also looking at your error it seems to be that **_your Nvida driver is too old_** NOT that the cuda version is incompatible.

**Reply by dmarsh400** [reply]
same issue, not sure exactly why this is happening, I'm assuming I will have to do a fresh install at this point, I will back up home and try that and see if that solves the issue.

**Reply by DrewThomasson** [reply]
I'm telling you man, I think you just need to update your Nvidia driver 🤷‍♂️

Not modifying your CUDA version

**Reply by DrewThomasson** [reply]
#### On the bright side,

Worst case scenario it should just automatically fall back on the cpu with that error.

So the program will probably be able to run anyway, with you ignoring the error.

It'll run, just albe slower idk 🤷‍♂️

**Reply by dmarsh400** [reply]
nvcc --version output

nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2021 NVIDIA Corporation
Built on Mon_Oct_11_21:27:02_PDT_2021
Cuda compilation tools, release 11.4, V11.4.152
Build cuda_11.4.r11.4/compiler.30521435_0

I am not sure if I can update the driver any more because I think I am at the highest version for the K80

nvidia-smi output

NVIDIA-SMI 470.256.02   Driver Version: 470.256.02   CUDA Version: 11.4

**Reply by DrewThomasson** [reply]
Well it looks like your CUDA version is still 11.4 and not 11.8

### BUT

but also if your getting the error with 11.4 then it might mean that **K80 might not even support CUDA 11.4**


Ill go test out seeing if VoxNovel can run with like CUDA 10 version get back to you with my results

**Reply by dmarsh400** [reply]
that would be great because I think you're right about k80 not being supported, some docs say it is and some say its deprecated and may not work

**Reply by DrewThomasson** [reply]
yeah you may have to use cuda 10

https://forums.developer.nvidia.com/t/nvidia-tesla-k80-cuda-version-support/67676

I'll get back with my VoxNovel results on if it even runs on CUDA 10 or not

**Reply by DrewThomasson** [reply]
Seems I'm unable to get CUDA 10 installed on google coal to test this out.

You'll have to just change the CUDA version on your own,

It looks like your running this in a wsl env right?

**Reply by dmarsh400** [reply]
no I am running ubuntu 22.04, looks like I may have to downgrade to 18.04 to get Cuda 10.2 installed

**Reply by DrewThomasson** [reply]
Dang,

I mean I guess you wouldn't of been able to use CUDA acceleration on any other projects anyway without doing this

**Reply by DrewThomasson** [reply]
# Testing Results: WSL Environment with Ubuntu 18.04

I downgraded and installed CUDA 10.2 using a custom script I created here:

- [Downgrade CUDA 10.2 for Ubuntu 18.04](https://github.com/DrewThomasson/WSL-scripts/blob/main/Downgrade-CUDA-10.2-Ubuntu18.04)

Next, I ran the auto-install script as normal:

```bash
yes | wget -O - https://raw.githubusercontent.com/DrewThomasson/VoxNovel/main/shell_install_scripts/Ubuntu-install.sh | bash
```

I then had to manually re-download the Xtts model files using this script:

- [Download Default Xtts V2 Model Files](https://github.com/DrewThomasson/WSL-scripts/blob/main/Download-Default_xttsv2-Model-Files.sh)

Afterward, I was able to run VoxNovel with the following caveats:

## Results:

- **BookNLP** works with CUDA acceleration.
- **Xtts** does not seem to use CUDA acceleration and defaults to the CPU. 
    - You may get different results on your system since this test was run in a Windows WSL environment. 

Feel free to double-check on your end!

**Reply by dmarsh400** [reply]
I was able to get CUDA 11.8 working with epub2txt and it is using the K80 GPU just fine.

After I finish this last batch of audio books I will attempt to install VoxNovel and see if it will also use GPU as in the previous post you said it should work fine with 11.8

the last time I installed VoxNovel it seemed to break epub2txt for some reason, that is why I am going to finish these books and then install VoxNovel to redo all the books so that way I have a backup to listen to while driving.

Anything I need to look out for while installing? I already have pytorch 2.1.0 I believe so that should not be an issue.

**Reply by DrewThomasson** [reply]
Not that I know of on the top of my head...

- But you say that it broke the epub2txt pip package?
- That pip package isn't even needed lol, it's a backup for if calibre is NOT installed, heres a modified gui_run.py file with it removed just in case.
[VoxNovel-no-epub2txt.py.zip](https://github.com/user-attachments/files/17216203/VoxNovel-no-epub2txt.py.zip)



I guess just check if you can get XTTS running on its own, cause thats theist and default TTS backend that VoxNovel uses.

[xtts github](https://github.com/coqui-ai/TTS)

line used for it after a 
`pip install tts==0.21.3`

Synthesizing speech by 🐸XTTS

🐍 Python API

```python
import torch
from TTS.api import TTS

# Get device
device = "cuda" if torch.cuda.is_available() else "cpu"

# List available 🐸TTS models
print(TTS().list_models())

# Init TTS
tts = TTS("tts_models/multilingual/multi-dataset/xtts_v2").to(device)

# Run TTS
# ❗ Since this model is multi-lingual voice cloning model, we must set the target speaker_wav and language
# Text to speech list of amplitude values as output
wav = tts.tts(text="Hello world!", speaker_wav="my/cloning/audio.wav", language="en")
# Text to speech to a file
tts.tts_to_file(text="Hello world!", speaker_wav="my/cloning/audio.wav", language="en", file_path="output.wav")
```

**Reply by dmarsh400** [reply]
No sorry I mean it broke the epub2tts which is another program I was using for text to audiobook. 

I think it uses a different version of some dependency so when I install VoxNovel it replaces something.

Anyway I'm pretty confident I will get it working now that I know a lot more about the way CUDA and Pytorch work.

Ill keep  you updated.

Also I had a question, I noticed when I ran it before that most of the voices are British sounding, is there somewhere I can get samples of the voices and is there a way to port the voices from epub2tts into VoxNovel or are they already present?

I haven't played around with VoxNovel enough to find the range of voices but a book with say 30 unique characters it would be nice to have enough voices to supply each one with their own.

**Reply by DrewThomasson** [reply]
## voxnovel voice clone button

Yes there is a clone voice button in the VoxNovel gui actually, just give it a like 15-30 second audio sample of any person talking, give the name of the voice actor and select the gender for auto selection, and then that voice will be added to the list of available voices actors.

It should walk you through the process anyway

![image](https://github.com/user-attachments/assets/746b7354-8b24-4afb-87d8-3d29a0e52ff9)

## About getting epub2tts voices

Also epub2tts seems to also use XTTS as the TTS engine, so it might have a bunch of voice samples somewhere in its codebase so you can probs just easily grab voices from there.


## when selecting one of the voices from the dropdown of VoxNovel for a character 

- it should attempt to play one of the audio samples of that voice actor locally stored in the "tortoise" folder

**Reply by dmarsh400** [reply]
Perfect, I will play around with it and see what I can do.

Thanks for all the help!

**Reply by DrewThomasson** [reply]
Oh, also why not install miniconda and just set up two different conda python envs  with different pips for both programs?

**Comment by DrewThomasson** [comment]
Testing on my end in a Google Colab right now...

**Comment by DrewThomasson** [comment]
I mean it work in Google Colab on my end with CUDA 11.8

<img width="656" alt="image" src="https://github.com/user-attachments/assets/f436f7e4-11c5-4d15-a603-29533a9d1fd4">


See full run details in this ipynb colab file 

[Colab_Headless_VoxNovel.ipynb.zip](https://github.com/user-attachments/files/16897590/Colab_Headless_VoxNovel.ipynb.zip)

**Reply by DrewThomasson** [reply]
Update: may of given the wrong colab file here is if it was wrong 
[Copy_of_Colab_Headless_VoxNovel.ipynb.zip](https://github.com/user-attachments/files/16897768/Copy_of_Colab_Headless_VoxNovel.ipynb.zip)

**Comment by dmarsh400** [comment]
I didn’t even know that was an option, I’ll have to look into that.I’m kind of new to this whole thing but I’ve been running Linux exclusively for 17 years On Oct 1, 2024, at 16:23, Drew Thomasson ***@***.***> wrote:﻿
Oh, also why not install miniconda and just set up two different conda python envs  with different pips for both programs?

**Reply by DrewThomasson** [reply]
lol yeah, 

## You might already have it installed actually from VoxNovel
also if you used the auto-installer command for VoxNovel then you probs already have conda installed, cause that should be doing that, 

try typing in 

`conda env list`

and it should show the the conda envs that exist

to activate one "start using one" type `conda activate envName`

to create one you can do something like this for instance

`conda create --name {yourCustomeEnvName} python={pythonNUMBER}`

### Example for creating a conda env named booknlp that uses python 3.7 and how to activate it:

```bash
conda create --name booknlp python=3.7
conda activate booknlp
#to deactivate type:
conda deactivate
```

once your in a activate conda env then you could just be able to do normal pip install commands tbh and it'll make those pip installs only apply to the conda env your currently in

lol Yeah, miniconda makes it a LOT easer handling python projects that use multiple python versions or have diffrent conflicting dependencies etc

**Comment by dmarsh400** [comment]
Perfect, I will do this tomorrow, if I can’t figure something out I may hit you up for more advice if you don’t mind.On Oct 1, 2024, at 20:14, Drew Thomasson ***@***.***> wrote:﻿
lol yeah,
You might already have it installed actually from VoxNovel
also if you used the auto-installer command for VoxNovel then you probs already have conda installed, cause that should be doing that,
try typing in
conda env list
and it should show the the conda envs that exist
to activate one "start using one" type conda activate envName
to create one you can do something like this for instance
conda create --name {yourCustomeEnvName} python={pythonNUMBER}
Example for creating a conda env named booknlp that uses python 3.7 and how to activate it:
conda create --name booknlp python=3.7
conda activate booknlp
#to deactivate type:
conda deactivate

**Comment by DrewThomasson** [comment]
Sounds like a plan 👍

**Comment by dmarsh400** [comment]
so Im running into the same issue, I installed VoxNovel thinking it would keep my version of Pytorch "2.1.0 cu118" but it uninstalled that version and installed some other version and now it will not run the GPU, its saying again

 UserWarning: CUDA initialization: The NVIDIA driver on your system is too old (found version 11040). Please update your GPU driver by downloading and installing a new version from the URL: http://www.nvidia.com/Download/index.aspx Alternatively, go to: https://pytorch.org to install a PyTorch version that has been compiled with your version of the CUDA driver. (Triggered internally at ../c10/cuda/CUDAFunctions.cpp:108.)
  return torch._C._cuda_getDeviceCount() > 0

when I attempt to install the proper version it gives this error

I tried both PIP and Conda to install ill add both below

pip install torch==2.1.0 torchvision==0.16.0 torchaudio==2.1.0 --index-url https://download.pytorch.org/whl/cu118
Looking in indexes: https://download.pytorch.org/whl/cu118
ERROR: Could not find a version that satisfies the requirement torch==2.1.0 (from versions: 2.2.0+cu118, 2.2.1+cu118, 2.2.2+cu118, 2.3.0+cu118, 2.3.1+cu118, 2.4.0+cu118, 2.4.1+cu118)
ERROR: No matching distribution found for torch==2.1.0

conda install pytorch==2.1.0 torchvision==0.16.0 torchaudio==2.1.0 pytorch-cuda=11.8 -c pytorch -c nvidia
Channels:
 - pytorch
 - nvidia
 - defaults
Platform: linux-64
Collecting package metadata (repodata.json): done
Solving environment: / warning  libmamba Added empty dependency for problem type SOLVER_RULE_UPDATE
failed

LibMambaUnsatisfiableError: Encountered problems while solving:
  - package torchvision-0.16.0-py310_cpu requires python >=3.10,<3.11.0a0, but none of the providers can be installed

Could not solve for environment specs
The following packages are incompatible
├─ pin-1 is installable and it requires
│  └─ python 3.12.* , which can be installed;
└─ torchvision 0.16.0  is not installable because there are no viable options
   ├─ torchvision 0.16.0 would require
   │  └─ python >=3.10,<3.11.0a0 , which conflicts with any installable versions previously reported;
   ├─ torchvision 0.16.0 would require
   │  └─ python >=3.11,<3.12.0a0 , which conflicts with any installable versions previously reported;
   ├─ torchvision 0.16.0 would require
   │  └─ python >=3.8,<3.9.0a0 , which conflicts with any installable versions previously reported;
   └─ torchvision 0.16.0 would require
      └─ python >=3.9,<3.10.0a0 , which conflicts with any installable versions previously reported.


not sure what I'm doing wrong

**Reply by dmarsh400** [reply]
I am able to install pytorch 2.4.0 cu118 but im still getting the error that the driver is too old and to compile a version of pytorch for my driver.

Im going to try pytorch 2.4.1 cu118

**Reply by dmarsh400** [reply]
installed pytorch 2.4.1cu118 with the same results, here is a copy of my environment

<frozen runpy>:128: RuntimeWarning: 'torch.utils.collect_env' found in sys.modules after import of package 'torch.utils', but prior to execution of 'torch.utils.collect_env'; this may result in unpredictable behaviour
Collecting environment information...
PyTorch version: 2.4.1
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.25.0
Libc version: glibc-2.35

Python version: 3.12.4 | packaged by Anaconda, Inc. | (main, Jun 18 2024, 15:12:24) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.8.0-45-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: Tesla K80
GPU 1: Tesla K80

Nvidia driver version: 470.256.02
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        43 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               4
On-line CPU(s) list:                  0-3
Vendor ID:                            AuthenticAMD
Model name:                           AMD Ryzen 3 3200G with Radeon Vega Graphics
CPU family:                           23
Model:                                24
Thread(s) per core:                   1
Core(s) per socket:                   4
Socket(s):                            1
Stepping:                             1
Frequency boost:                      enabled
CPU max MHz:                          3600.0000
CPU min MHz:                          1400.0000
BogoMIPS:                             7187.00
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb hw_pstate ssbd ibpb vmmcall fsgsbase bmi1 avx2 smep bmi2 rdseed adx smap clflushopt sha_ni xsaveopt xsavec xgetbv1 clzero irperf xsaveerptr arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif overflow_recov succor smca sev sev_es
L1d cache:                            128 KiB (4 instances)
L1i cache:                            256 KiB (4 instances)
L2 cache:                             2 MiB (4 instances)
L3 cache:                             4 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-3
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Mitigation; untrained return thunk; SMT disabled
Vulnerability Spec rstack overflow:   Mitigation; SMT disabled
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Retpolines; IBPB conditional; STIBP disabled; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.3
[pip3] torch==2.4.1
[pip3] torchaudio==2.4.1
[pip3] torchvision==0.19.0+cu118
[pip3] triton==3.0.0
[conda] blas                      1.0                         mkl  
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] libjpeg-turbo             2.0.0                h9bf148f_0    pytorch
[conda] mkl                       2023.1.0         h213fc3f_46344  
[conda] mkl-service               2.4.0           py312h5eee18b_1  
[conda] mkl_fft                   1.3.10          py312h5eee18b_0  
[conda] mkl_random                1.2.7           py312h526ad5a_0  
[conda] numpy                     1.26.3                   pypi_0    pypi
[conda] numpy-base                1.26.4          py312h0da6c21_0  
[conda] pytorch                   2.4.1           py3.12_cuda11.8_cudnn9.1.0_0    pytorch
[conda] pytorch-cuda              11.8                 h7e8668a_6    pytorch-test
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torch                     2.4.0+cu118              pypi_0    pypi
[conda] torchaudio                2.4.0+cu118              pypi_0    pypi
[conda] torchtriton               3.0.0                     py312    pytorch
[conda] torchvision               0.19.0+cu118             pypi_0    pypi
[conda] triton                    3.0.0                    pypi_0    pypi

**Reply by DrewThomasson** [reply]
lol, make sure to use the multiple conda envs to make it easier for you to organize and such

- one for VoxNovel
- one for epub2tts whatnot 

also when in a conda env or in general whatnot you can run the command
```bash
pip freeze > requirements.txt
```
to put all of the pips and exact versions in a requirements.txt file

to later install with a

```bash
pip install -r requirements.txt --no-deps
```

command

might make it easier if you ever want to share the pip env with anyone if they need to replicate what you did,

**Reply by DrewThomasson** [reply]
Tell me if you ever get it working tho!
:)

**Reply by dmarsh400** [reply, marked as answer]
I got it to work, its using CUDA and my GPU

I did what you said and made a separate conda env.

then I did a manual install and resolved dependencies manually using the ubuntu_requirements.txt and resolving dependencies as they were shown in the error messages.

**Reply by DrewThomasson** [reply]
nice nice!

**Reply by DrewThomasson** [reply]
I guess if its working now and all you needed was the pip changes 

If you do 

```bash
pip freeze > requirements.txt
```

and paste that into this thread for anyone else that runs into your issue so then they can just run a 

```bash
pip install --no-deps -r requirements.txt
```

to replicate all your pips in your conda env?🥺

**Reply by dmarsh400** [reply]
Where do i find the requirements.txt file so I can paste it?

**Reply by DrewThomasson** [reply]
running 
```bash
pip freeze > requirements.txt
```

should just create a requirements.txt file in the local dir you ran that command in

you also could give it anyname lol, doesnt have to be named requirements.txt,
could be like CUDA11.8_k80_Ubuntu_requirments.txt or something idk

for instance

```bash
pip freeze > CUDA11.8_k80_Ubuntu_requirments.txt
```

**Reply by dmarsh400** [reply]
[CUDA11.8_K80_Ubuntu_requirements.txt](https://github.com/user-attachments/files/17236372/CUDA11.8_K80_Ubuntu_requirements.txt)

**Reply by DrewThomasson** [reply]
thxthxthx 

- Now anyone should be able to replicate your work when they also run into this issue
